### PR TITLE
Update registration system and enhance related functionalities

### PR DESCRIPTION
**Pull request by Pawel-Mackiewicz**
### Description
This pull request includes major updates and enhancements to the registration system and associated features:

- **Registration system is now available via API**
- **Demo user registration**: Added `/api/registration/demo` endpoint along with `DemoRegistrationService` and validation improvements.
- **Exception handling**: Introduced more granular exceptions for uniqueness validations like phone, email, and PESEL.
- **Integration tests**: Added comprehensive integration tests for the Registration System.
- **Service layer refactor**: Redesigned registration system architecture, introducing proper service interfaces, bonus grants for new users, and better modularity.
- **Logging and maintenance**: Enhanced logging configurability, documentation, and exception handling for improved maintainability.
- **Code reorganization**: Reorganized user services, email-related classes, and test structures for better clarity and scalability.
- **Updated Maven and versioning**: Upgraded Maven image to 3.9.9 and finalized release version 0.4.7.

Lastly, the changes include refactoring for clarity (e.g., renaming `Email` to `EmailAddress`), improving locking mechanisms, and restricting workflow triggers in CI to the main branch. 

This update ensures a robust registration system with improved functionality and better compliance with coding standards.

## Summary by CodeRabbit

- **New Features**
  - Introduced comprehensive user registration system with REST API endpoints for regular and demo users, including DTOs, mapping, validation, and welcome bonus granting.
  - Added new exception types for detailed error handling (duplicated email, PESEL, phone number, username, user field null, username validation).
  - Added configurable welcome bonus amount and default bank account IBAN.
  - Added stack trace logging toggle via configuration.
  - Added utility to calculate elapsed time from start of month for request IDs.
  - Added MDC context logging for account locking operations.
  - Added `BankAccountProvider` component for default bank account access.
- **Improvements**
  - Unified email value object usage to `EmailAddress` across the entire application.
  - Enhanced logging with MDC context for improved traceability.
  - Allowed unauthenticated access to registration API endpoints.
  - Refined user creation, validation, and password handling for robustness.
  - Centralized logging patterns and improved error message formatting.
  - Reorganized package structure and imports for clarity.
  - Improved validation service to enforce non-null fields upfront and throw more specific exceptions.
  - Simplified locking backoff delay calculation and logging.
  - Updated username generation to throw specific exception on invalid input.
- **Bug Fixes**
  - Fixed validation and error handling for user registration edge cases including duplicate fields and invalid inputs.
- **Documentation**
  - Enhanced API documentation with OpenAPI schema annotations and example values for error and validation responses.
- **Tests**
  - Added extensive integration and unit tests covering registration workflow, bonus granting, validation, and error scenarios.
  - Introduced test utilities for generating valid and invalid registration requests.
  - Added integration tests for registration controller and service covering success and failure cases.
  - Added unit tests for bonus granting and registration services.
- **Chores**
  - Upgraded dependencies and build configurations.
  - Refactored and reorganized code structure for clarity and maintainability.
  - Updated configuration files for consistency and new features.
  - Cleaned and consolidated import statements across tests and main code.

### CODE REVIEW

File: src/main/java/info/mackiewicz/bankapp/system/registration/dto/demo/RegistrationRequestFactory.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        request.setPassword(password);
+        request.setConfirmPassword(password);
+        request.setFirstname(DEFAULT_FIRSTNAME);
+        request.setLastname(DEFAULT_LASTNAME);
+        request.setDateOfBirth(DEFAULT_DATE_OF_BIRTH);
+        request.setPesel(generateRandomPesel());
+        request.setPhoneNumber(generateRandomPhoneNumber());
+
+        return request;
+    }
```
[the remark this answers is not in the document]
Comment (Pawel-Mackiewicz) [reply src/main/java/info/mackiewicz/bankapp/system/registration/dto/demo/RegistrationRequestFactory.java:40]: @coderabbitai 
is only @NonNull enough?

File: src/main/java/info/mackiewicz/bankapp/shared/util/BankAccountProvider.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    private Iban defaultBankIban;
+
+    @PostConstruct
+    private void init() {defaultBankIban = Iban.valueOf(defaultBankIbanString);}
+
+    private final AccountService accountService;
+
+    public Account getBankAccount() {
+        return accountService.getAccountByIban(defaultBankIban);
+    }
```
[the remark this answers is not in the document]
Comment (Pawel-Mackiewicz) [reply src/main/java/info/mackiewicz/bankapp/shared/util/BankAccountProvider.java:33]: bank account couldn't be null. ever.

File: src/main/java/info/mackiewicz/bankapp/system/registration/dto/RegistrationMapperImpl.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        return User.builder()
+                .withFirstname(request.getFirstname())
+                .withLastname(request.getLastname())
+                .withPesel(request.getPesel())
+                .withDateOfBirth(request.getDateOfBirth())
+                .withEmail(request.getEmail())
+                .withPhoneNumber(request.getPhoneNumber())
+                .withPassword(request.getPassword())
+                .build();
+    }
```
[the remark this answers is not in the document]
Comment (Pawel-Mackiewicz) [reply src/main/java/info/mackiewicz/bankapp/system/registration/dto/RegistrationMapperImpl.java]: @coderabbitai
added NonNull annotations.

File: src/main/java/info/mackiewicz/bankapp/shared/web/dto/BaseApiError.java [hunk 4]
```diff
@@ -32,13 +33,16 @@
...
+    @Schema(example = "VALIDATION_ERROR")
+    private final String title;
     
     /** Detailed description of the error */
-    private String message;
+    @Schema(example = "Validation failed. Please check your input and try again.")
+    private final String message;
     
     /** The URI path where the error occurred */
     private String path;
```
[the remark this answers is not in the document]
Comment (Pawel-Mackiewicz) [reply src/main/java/info/mackiewicz/bankapp/shared/web/dto/BaseApiError.java]: @coderabbitai
done

File: src/main/java/info/mackiewicz/bankapp/system/registration/controller/RegistrationController.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+                                              "status": "CONFLICT",
+                                              "title": "USER_ALREADY_EXISTS",
+                                              "message": "User with these credentials already exists.",
+                                              "path": "/api/registration/regular",
+                                              "timestamp": "11-04-2025 16:18:29"
+                                            }
+                                            """)))
+    })
+    @PostMapping("/regular")
+    ResponseEntity<RegistrationResponse> registerUser(RegistrationRequest request);
```
[the remark this answers is not in the document]
Comment (Pawel-Mackiewicz) [reply src/main/java/info/mackiewicz/bankapp/system/registration/controller/RegistrationController.java]: @coderabbitai
annotations added

File: src/main/java/info/mackiewicz/bankapp/system/registration/controller/RegistrationController.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+                                              "status": "CONFLICT",
+                                              "title": "USER_ALREADY_EXISTS",
+                                              "message": "User with these credentials already exists.",
+                                              "path": "/api/registration/demo",
+                                              "timestamp": "11-04-2025 16:18:29"
+                                            }
+                                            """)))
+    })
+    @PostMapping("/demo")
+    ResponseEntity<RegistrationResponse> registerDemoUser(DemoRegistrationRequest request);
```
[the remark this answers is not in the document]
Comment (Pawel-Mackiewicz) [reply src/main/java/info/mackiewicz/bankapp/system/registration/controller/RegistrationController.java]: @coderabbitai
added

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -2,33 +2,51 @@ name: Java CI with Maven
...
-      with:
-        name: test-results
-        path: target/surefire-reports/
-    - name: Upload Coverage Report
-      uses: actions/upload-artifact@v4
-      with:
-        name: coverage-report
-        path: target/site/jacoco/
\ No newline at end of file
+      - uses: actions/checkout@v3
```
[the remark this answers is not in the document]
Comment (Pawel-Mackiewicz) [reply .github/workflows/ci.yml]: @coderabbitai
done
